### ISSUE TITLE
`nearest_n_within` does not limit the number of items when not sorted

### ISSUE THREAD

**Issue by Nahor** [sdd-org/kiddo#168]
In kiddo `v4.2.0`, `nearest_n_within` will only limit the number of items if sorting is enabled. If it's disabled, all the items within the radius are returned.

E.g. in the `cities` example, replace `nearest_n` with `nearest_n_within` and a radius of 10km. 
```diff
-let nearest_5_idx = kdtree.nearest_n::<SquaredEuclidean>(&query, 5);
+let nearest_5_idx = kdtree.nearest_n_within::<SquaredEuclidean>(
+     &query,
+     kilometres_to_unit_sphere_squared_euclidean(10.0),
+    5,
+    false,
+);
```
This outputs 13 cities (for my data set) instead of the expected 5. If the list is sorted, then it works as expected.

---
Also note that the example in the documentation for `nearest_n_within` creates a tree of 2 items, asks for 2 items, and then is happy because it got 2 items in return!

**Comment by sdd** [comment]
Whoops! Thanks for spotting this. Will fix in an upcoming release as soon as I can.

**Comment by ezrasingh** [comment]
Do you have any information on the release date for the next version? 

I encountered the same bug in my [project](https://github.com/ezrasingh/geoprox/blob/6385ad9840ef4f8ade6c1ebc2fdedd7d5974ef4d/geoprox-core/src/cache.rs#L162) and was wondering when the fix from #175 will be available.

**Comment by Nahor** [comment]
@ezrasingh , you can always use the git repo instead of a released crate:
```
kiddo = { git = "https://github.com/sdd/kiddo.git", rev = "cc62c6d76accdeba75c0f55db80c28eab8a21ae7" }
```

**Comment by ezrasingh** [comment]
Hey @Nahor,

I [pinned Kiddo to that revision in my crate](https://github.com/ezrasingh/geoprox/commit/0f0c2a48180e9fd5232d990a6365dd589d773782), but my unit tests are still failing. Initially, I thought it might be due to stale code on my machine. However, the tests also [failed in CI](https://github.com/ezrasingh/geoprox/actions/runs/10310781976/job/28543157407#step:5:123).

Could this possibly be a regression?

**Comment by ezrasingh** [comment]
I just submitted a PR (#180) to validate the consistency of the behavior when `max_qty` = 0. During testing, I found that the function handles this case differently depending on whether the results are sorted or not. Specifically, the issue arises when `sorted` = false, where the function does not return an empty vec as expected, unlike when `sorted` = true.

This difference explains why pinning Kiddo did not work for my case. While I can work around this with a conditional branch, I don't believe this inconsistent behavior is intended. If it works with sorted results, it should also work with unsorted results, ensuring consistent handling of `max_qty` = 0.

**Comment by sdd** [comment]
Hey all - just released v4.2.1: https://crates.io/crates/kiddo/4.2.1

**Comment by ezrasingh** [comment]
Hi @sdd, so far [v4.2.1](https://github.com/sdd/kiddo/releases/tag/v4.2.1) has been great however I notice this release still fails one of the edge case I added in PR #180:

Calling `nearest_n_within` when `max_qty` = 0:
- using `sorted` = `false`  will panic
- using `sorted` = `true` returns an empty `Vec` as expected

**Comment by sdd** [comment]
Aah hey Ezra, I'll take a look over this weekend and see if I can get a fix out.

**Comment by sdd** [comment]
This issue should be resolved in https://github.com/sdd/kiddo/pull/186 as part of the rewrite of the `ImmutableKdTree`, which was just released as Kiddo v5.0.0 🎉 

https://crates.io/crates/kiddo/5.0.0

I know this issue was raised a while ago, and you have probably moved on, but if you do get chance to try v5 I'd love to hear any feedback. I'm closing this issue.

**Comment by ezrasingh** [comment]
> This issue should be resolved in [#186](https://github.com/sdd/kiddo/pull/186) as part of the rewrite of the `ImmutableKdTree`, which was just released as Kiddo v5.0.0 🎉
> 
> https://crates.io/crates/kiddo/5.0.0
> 
> I know this issue was raised a while ago, and you have probably moved on, but if you do get chance to try v5 I'd love to hear any feedback. I'm closing this issue.

I tried v5, and not only does it handle the edge case, but it also outperforms my previous benchmarks 👍🏾 great work @sdd

**Comment by sdd** [comment]
Thank you @ezrasingh! Glad to hear it 😁
